### PR TITLE
Refactor Authentication Providers

### PR DESCRIPTION
**Pull request by jonathangreen**
## Description

I've been working on updating how we are storing configuration settings in the database. As part of that work, I started replacing the configuration settings for our authentication providers and realized that these classes need a bit of refactoring. So I broke that out into this separate PR.

## Motivation and Context

This PR does some general refactoring of the various Authentication classes in the CM code. 

- Generally I tried to add type hinting to as much of this code as I could, to assist with refactoring. 
  - Once I had accurate type hints, it was pretty obvious that some of the methods could return way too many types. It was pretty common to see: `PatonData | Patron | ProblemDetail | None | bool` 😓
  - So some of the refactoring focused on reducing the number of types each method had to handle
- I also tried to update places where we were testing for none with `== None` to `is None`. 

### `Authenticator`

- Moved creating the logging into the constructor, so I could use the class dunder properties to name the logger.
https://github.com/ThePalaceProject/circulation/blob/9d39cc4e755b8823d7b3d4226f0c9463b266d4aa/api/authenticator.py#L540-L543
- Switched to using the logging context manager instead of the decorator, to use the new `self.log` instance property.

### `AuthenticationProvider`

- This base class is where the bulk of the changes happened. When I was working through the configuration settings, and trying to trace out which subclasses used which settings, it became evident that the class structure here did not align with how it was being used. 
- The `LIBRARY_SETTINGS` that are defined on this class, and all the associated constants were only being used by instances of `BasicAuthenticationProvider`, so it didn't make a lot of sense to have them defined on the base class.
- Most of the code in this class was being used to enforce the library identifier restriction, which is only used by `BasicAuthenticationProvider` as well.
- After these changes `AuthenticationProvider` becomes a very basic abstract base class that defines two abstract methods: `authenticated_patron` and `get_credential_from_header`. 
- Almost all of the code that was removed from `AuthenticationProvider` was directly moved into `BasicAuthenticationProvider`.

### `BasicAuthenticationProvider`

- There are two sets of changes for `BasicAuthenticationProvider`
  1. The code that was moved from `AuthenticationProvider`
  2. Refactoring to reduce the excessive number of types that each method had to handle.
- After the changes in this PR `BasicAuthenticationProvider` becomes an abstract base class with two abstract methods `remote_patron_lookup` and `remote_authenticate`.
- In order to simplify the logic here, the `_remote_patron_lookup` function was removed, and replaced with `remote_patron_lookup`. The public method used to just run `enforce_library_identifier_restriction`, and then call the private `_remote_patron_lookup` method. Digging through the logic, `enforce_library_identifier_restriction` was already being enforced elsewhere in the `authenticate` method, so this private method distinction was unnecessary.
- `remote_authenticate` was updated to that implementing classes return `None` instead of `False` or `None` on failure.
- Remove the `external_type_regular_expression` configuration setting entirely, since we are not using it in any of the 400 - 500 libraries we currently have configured.

### `api/firstbook2.py` & `api/kansas_patron.py`
Both these providers were just updated to implement the `remote_patron_lookup` abstract method, since the no-op implementation of this was removed from `BasicAuthenticationProvider` and made abstract.

### `api/millenium_patron.py`

- Switch from `False` to `None` as return type
- Additional `None` checks that were flagged as necessary by Mypy

### `api/saml/provider.py`

- Implement the abstract method `get_credential_from_header` since the no-op version was removed from the base class. This is the same as the code that used to be defined on the base class.

### `api/sip/__init__.py`

- Move the `INSTITUTION_ID` setting that was only used in this provider, from the base class into this provider.

### `api/sirsidynix_authentication_provider.py`

- Update `False` to `None` as per new conventions
- Some additional `None` checks, prompted by mypy.

## How Has This Been Tested?

- Ran unit tests locally and in CI
- Added unit tests for new code paths
- Done some basic, but not extensive testing locally in the admin UI

## Checklist

- [X] I have updated the documentation accordingly.
- [X] All new and existing tests passed.

### CODE REVIEW

File: api/authenticator.py [hunk 1]
```diff
@@ -1549,134 +1328,26 @@ def authenticated_patron(self, _db, header):
...
-        if PatronUtility.needs_external_sync(patron):
-            self.update_patron_metadata(patron)
-        if patron.cached_neighborhood and not patron.neighborhood:
-            # Patron.neighborhood (which is not a model field) was not
-            # set, probably because we avoided an expensive metadata
-            # update. But we have a cached_neighborhood (which _is_ a
-            # model field) to use in situations like this.
-            patron.neighborhood = patron.cached_neighborhood
-        return patron
+    @abstractmethod
```
Comment (io7m) [inline api/authenticator.py:1333]: :+1:

File: api/authenticator.py [hunk 2]
```diff
@@ -1955,43 +1718,84 @@ def __init__(self, library, integration, analytics=None):
...
-            return patron_info
+            if isinstance(restriction, str):
+                self.library_identifier_restriction = restriction.strip()
+            else:
+                self.library_identifier_restriction = restriction
+
+    @abstractmethod
+    def remote_patron_lookup(
+        self, patron_or_patrondata: PatronData | Patron
+    ) -> PatronData | ProblemDetail | None:
```
[the remark this answers is not in the document]
Comment (jonathangreen) [reply api/authenticator.py:1771]: @io7m I updated the comment here to add more detail about how this function is expected to be implemented, and the return types.

[the remark this answers is not in the document]
Comment (io7m) [reply api/authenticator.py:1770]: That's a bit better.

**Review by io7m — APPROVED** [review]
This looks good to me!
Always happy to see more types, and I'm looking forward to what this does for configuration settings.
